pkg/beacon/relay/groupselection/groupselection_test.go

```diff
@@ -14,180 +18,385 @@ import (
+				576460752303423488,
 			},
-			expectedSubmittedTicketsCount: 2,
 		},
 	}
 
-	for _, test := range tests {
-		chainConfig := &config.Chain{
-			GroupSize:               test.groupSize,
-			TicketSubmissionTimeout: 6,
-		}
-
-		chain := &stubGroupInterface{
-			groupSize: test.groupSize,
-		}
-
-		blockCounter, err := local.BlockCounter()
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		onGroupSelected := func(result *Result) {}
-
-		err = startTicketSubmission(
-			test.initialSubmissionTickets,
-			test.reactiveSubmissionTickets,
-			chain,
-			blockCounter,
-			chainConfig,
-			0, // start block height
-			onGroupSelected,
-		)
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		err = blockCounter.WaitForBlockHeight(
-			chainConfig.TicketSubmissionTimeout,
-		)
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		submittedTicketsCount, err := chain.GetSubmittedTicketsCount()
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		expectedCount := big.NewInt(int64(test.expectedSubmittedTicketsCount))
-		if expectedCount.Cmp(submittedTicketsCount) != 0 {
-			t.Fatalf(
-				"unexpected number of submitted tickets\nexpected: [%v]\nactual:   [%v]",
-				expectedCount,
-				submittedTicketsCount,
-			)
-		}
+	for testName, test := range tests {
+		t.Run(testName, func(t *testing.T) {
+			existingChainTickets := make([]*chain.Ticket, 0)
+			for _, existingChainTicket := range test.existingChainTickets {
+				chainTicket, err := toChainTicket(
+					newTestTicket(0, existingChainTicket),
+				)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				existingChainTickets = append(existingChainTickets, chainTicket)
+			}
+
+			relayChain := &stubGroupInterface{
+				groupSize:        groupSize,
+				submittedTickets: existingChainTickets,
+			}
+
+			for roundIndex := uint64(0); roundIndex <= ticketSubmissionRounds; roundIndex++ {
+				roundLeadingZeros := ticketSubmissionRounds - roundIndex
+
+				candidateTickets, err := roundCandidateTickets(
+					relayChain,
+					tickets,
+					roundIndex,
+					roundLeadingZeros,
+					groupSize,
+				)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if !reflect.DeepEqual(
+					test.expectedCandidateTicketsPerRound[roundIndex],
+					candidateTickets,
+				) {
+					t.Fatalf(
+						"unexpected candidate tickets for round [%v]\n"+
+							"expected: [%v]\nactual:   [%v]",
+						roundIndex,
+						test.expectedCandidateTicketsPerRound[roundIndex],
+						candidateTickets,
+					)
+				}
+
+				for _, ticket := range candidateTickets {
+					chainTicket, err := toChainTicket(ticket)
+					if err != nil {
+						t.Fatal(err)
+					}
+
+					relayChain.SubmitTicket(chainTicket)
+				}
+			}
+
+			chainTickets, err := relayChain.GetSubmittedTickets()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if !reflect.DeepEqual(test.expectedChainTickets, chainTickets) {
+				t.Fatalf(
+					"unexpected chain tickets\n"+
+						"expected: [%v]\nactual:   [%v]",
+					test.expectedChainTickets,
+					chainTickets,
+				)
+			}
```

Comment:
I think we should get rid of this part as we are testing `toChainTicket`, `SubmitTicket` and `GetSubmittedTicktets` behaviour here. Checking which tickets and at which round were returned by `roundCandidateTickets` should be enough for this test.

Reply:
In the previous approach, we were testing the entire `startTicketSubmission`. That's why we have the entire `stubGroupInterface` here. I think we have two options:

1. Apply the comment above and write additional unit test stressing `startTicketSubmission` with just one case making sure tickets are submitted in correct rounds. We can have just 2-3 rounds for this test.
2. Make this test stress the entire `startTicketSubmission`. We'll probably have to [parameterize round duration](https://github.com/keep-network/keep-core/pull/1510#discussion_r402868814) to do not keep this test running forever. 

I am personally in favor of the first approach as it seems simpler. You are the expert here so you decide though :)

Reply:
Ok, done!